### PR TITLE
Add utility script to merge Cachi2 and Syft SBOMs

### PR DESCRIPTION
**Pull request by brunoapimentel**
The script performs a deduplication of reported components, favoring Cachi2's report whenever possible:
- Removes any Syft component that has a matching PURL with a Cachi2 component
- Removes Golang locally replaced modules from Syft
- Removes non-registry Pip and NPM dependencies reported by Syft

More details about the deduplication can be found [here](https://github.com/brunoapimentel/rhtap-cachi2-sbom-merge/blob/main/find-duplicates.ipynb).

This script is primarily meant to perform the merging of both SBOMs in the RHTAP build pipeline. The reason to implement it in Cachi2 is because of its growing complexity, which would be rather unmaintanable if written inline in a Tekton yaml file.

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [n/a] Docs updated (if applicable)
- [n/a] Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

**Comment by chmeliik** [comment]
> Removes Pip non-Pypi dependencies reported by Syft

What about npm non-registry dependencies?

File: utils/merge_syft_sbom.py
Comment (chmeliik) [inline utils/merge_syft_sbom.py:21]: Depending on the intended behavior, this may need more parentheses
```python
In [1]: False and False or True
Out[1]: True
```

File: utils/merge_syft_sbom.py
Comment (chmeliik) [inline utils/merge_syft_sbom.py:39]: Maybe checking for vcs_url | download_url qualifiers would be more robust and re-usable for npm?

File: utils/merge_syft_sbom.py
Comment (chmeliik) [inline utils/merge_syft_sbom.py:47]: This also omits the fragment (subpath), that's probably intentional too?

File: utils/merge_syft_sbom.py
Comment (chmeliik) [inline utils/merge_syft_sbom.py:60]: Just FYI: not all [purl types](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst) should lowercase the name. But those that cachi2 currently supports should.

Comment (chmeliik) [inline utils/merge_syft_sbom.py:60]: > We need to encode special characters in the version part,

Here we take the `version?qualifiers#subpath` part. Encoding that as a whole breaks the purl, but I guess Syft never uses any qualifiers or subpath?

File: utils/merge_syft_sbom.py
Comment (chmeliik) [inline utils/merge_syft_sbom.py:75]: Is it safe to de-duplicate by `name` only? While it's not possible to have two different versions of the same dependency in one requirements file, it's possible to have them across different req files or across two different packages

File: utils/merge_syft_sbom.py
Comment (chmeliik) [inline utils/merge_syft_sbom.py:114]: IMO this doesn't belong in a merge script but in cachi2 itself

File: utils/merge_syft_sbom.py
Comment (chmeliik) [inline utils/merge_syft_sbom.py:121]: should be len < 3

maybe also exit with non-zero

or use argparse :)

File: utils/merge_syft_sbom.py
Comment (chmeliik) [reply utils/merge_syft_sbom.py:60]: Hmm, we already make that assumption [here](https://github.com/containerbuildsystem/cachi2/pull/223#discussion_r1259607142) so this is consistent

File: utils/merge_syft_sbom.py
Comment (brunoapimentel) [reply utils/merge_syft_sbom.py:75]: Yeah, I thought about that. Since Cachi2 does not report the version for vcs/file dependencies, there's no way to accurately determine that it is exactly the same dependency. 

This won't be a problem if all packages are processed by Cachi2, since we'll get Cachi2's report of all versions of this dependency. The corner case I see is if one of the packages is not processed by Cachi2, or if it's a package lying somewhere in the image and Syft picks it up.

I think the alternative option here is to live with duplicates of Pip non-PiPY dependencies =/
Unless you have another idea on how to handle this.

Comment (chmeliik) [reply utils/merge_syft_sbom.py:75]: What Syft SBOMs will we be merging with this exactly? The one generated from the source directory? From the built image? Both?

Comment (brunoapimentel) [reply utils/merge_syft_sbom.py:75]: Both Syft-source and Syft-image SBOMs are already merged in the build pipeline nowadays. So we'll do another merge, and add Cachi2's SBOM there.

I added a Pip few examples [here](https://github.com/brunoapimentel/rhtap-cachi2-sbom-merge), but I still need to update the investigation as I did with Golang.

Comment (chmeliik) [reply utils/merge_syft_sbom.py:75]: Hmm, theoretically any dependencies unknown to cachi2 should only be able to get there from the base image. I'd say it's fine to drop the Syft duplicates by name in this case

Comment (brunoapimentel) [reply utils/merge_syft_sbom.py:75]: Ack. I will make our approach clearer in the docstrings.

File: utils/merge_syft_sbom.py
Comment (brunoapimentel) [reply utils/merge_syft_sbom.py:114]: Agreed. We will need to do some merging magic here regardless.

File: utils/merge_syft_sbom.py
Comment (brunoapimentel) [reply utils/merge_syft_sbom.py]: Fixed.

File: utils/merge_syft_sbom.py
Comment (brunoapimentel) [reply utils/merge_syft_sbom.py:1]: Added.

File: utils/merge_syft_sbom.py
Comment (brunoapimentel) [reply utils/merge_syft_sbom.py]: I changed the script so it will merge any `metadata.tools` if they're available. The second part should be solved by #232.

File: utils/merge_syft_sbom.py
Comment (brunoapimentel) [reply utils/merge_syft_sbom.py]: I keep forgeting Python has these types of magic. Changed to argparse.

File: utils/merge_syft_sbom.py
Comment (brunoapimentel) [reply utils/merge_syft_sbom.py:43]: It is intentional, since Syft does not set any qualifiers or subpaths (for Golang and Pip, at least)

File: utils/merge_syft_sbom.py
Comment (brunoapimentel) [reply utils/merge_syft_sbom.py]: I added more info on the docstrings to try to clear this up.

File: utils/merge_syft_sbom.py
Comment (brunoapimentel) [reply utils/merge_syft_sbom.py]: Good point. I did not do any research on how NPM dedup would look like, though. I'll look into this tomorrow, and I bet it add more dedup rules to this already long script :confused: 

We can revisit this topic then.

File: tests/unit/data/sboms/merged.bom.json
Comment (chmeliik) [inline tests/unit/data/sboms/merged.bom.json:20]: Observation: we prefer the Syft SBOM for everything except `components` (for those, we merge but prefer cachi2)

The merging is meant to be a temporary solution, right? Otherwise there's some more to think about IMO, like mismatched specVersion (note that 1.5 is already released)

File: tests/unit/data/sboms/cachi2.bom.json
Comment (chmeliik) [inline tests/unit/data/sboms/cachi2.bom.json:3]: Would be good to add more metadata to the cachi2 test SBOM

**Comment by ben-alkov** [comment]
More of an "I don't really have anything to add so I'll just talk" than a real comment :D

I have a deep fondness for certain Functional Programming principles. One can be summarized as "Single Entry, Single Exit"[1], f.e. this

```python
    if "purl" not in component:
        return component.get("name", "") + "@" + component.get("version", "")

    if "@" in component["purl"]:
        parts = component["purl"].split("@")
        return parts[0].lower() + "@" + quote_plus(parts[1])
    else:
        return component["purl"]
```

can be refactored into (roughly; there are some loose ends)

```python
    val = None

    if component.get("purl", None):
        if "@" in component["purl"]:
            parts = component["purl"].split("@")
            val = parts[0].lower() + "@" + quote_plus(parts[1])
        else:
            val = component["purl"]
    else:
        # No purl
        val = component.get("name", "") + "@" + component.get("version", "")

    return val
```

I feel like it usually ends up being a case of more logic vs. more `return`s

[1] Obviously, most languages enforce single *entry* by design, except under carefully controlled circumstances

**Comment by ben-alkov** [comment]
@brunoapimentel;

LGTM so far - will NPM be a part of this PR?

**Comment by brunoapimentel** [comment]
> @brunoapimentel;
> 
> LGTM so far - will NPM be a part of this PR?

Yes, I'll push some small changes later on to cover NPM.

**Comment by brunoapimentel** [comment]
> More of an "I don't really have anything to add so I'll just talk" than a real comment :D
> 
> I have a deep fondness for certain Functional Programming principles. One can be summarized as "Single Entry, Single Exit"[1], f.e. this
> 
> ```python
>     if "purl" not in component:
>         return component.get("name", "") + "@" + component.get("version", "")
> 
>     if "@" in component["purl"]:
>         parts = component["purl"].split("@")
>         return parts[0].lower() + "@" + quote_plus(parts[1])
>     else:
>         return component["purl"]
> ```
> 
> can be refactored into (roughly; there are some loose ends)
> 
> ```python
>     val = None
> 
>     if component.get("purl", None):
>         if "@" in component["purl"]:
>             parts = component["purl"].split("@")
>             val = parts[0].lower() + "@" + quote_plus(parts[1])
>         else:
>             val = component["purl"]
>     else:
>         # No purl
>         val = component.get("name", "") + "@" + component.get("version", "")
> 
>     return val
> ```

I personally tend to use returns more to avoid to have control structures, specially if there needs to be nesting.

> I feel like it usually ends up being a case of more logic vs. more `return`s

Yeah, I agree. I guess it is probably better when the principle is language enforced.

I am trying to use more functional principles, btw, specially when I'm writing Java :smile:

**Comment by chmeliik** [comment]
> I am trying to use more functional principles, btw, specially when I'm writing Java :smile:

Obligatory endorsement for Haskell  :heart: 

I couldn't resist and had to publish my Brainfuck interpreter that I made last year :grin: https://github.com/chmeliik/Brainfuck

**Comment by brunoapimentel** [comment]
> > I am trying to use more functional principles, btw, specially when I'm writing Java smile
> 
> Obligatory endorsement for Haskell heart
> 
> I couldn't resist and had to publish my Brainfuck interpreter that I made last year grin https://github.com/chmeliik/Brainfuck

#### Why?
Why not?

File: tests/unit/data/sboms/merged.bom.json
Comment (brunoapimentel) [reply tests/unit/data/sboms/merged.bom.json]: Yeah, I think adding Cachi2 components to Syft's SBOM makes more sense, since Syft is the default one now (that exists in all pipelines regardless of them being hermetic).

Syft is still using spec 1.4 even in the lastest release, though  =/

File: utils/merge_syft_sbom.py
Comment (brunoapimentel) [inline utils/merge_syft_sbom.py:31]: I guess that `bitbucket:` and `gitlab:` might also be valid in purls, but I still couldn't get a Syft SBOM with these type of dependencies in it.

File: tests/unit/data/sboms/merged.bom.json
Comment (chmeliik) [reply tests/unit/data/sboms/merged.bom.json]: Using 1.4 is fine, the problem would be if we bump cachi2's specVersion, use some new features but keep 1.4 in the merged SBOM

File: utils/merge_syft_sbom.py
Comment (brunoapimentel) [reply utils/merge_syft_sbom.py:32]: Ok, so NPM rule is added.

Troblem about defaulting to `vcs_url | download_url` is that we only check for Cachi2 non-registry components for Pip. For NPM, we check for Syft non-registry components, and for Golang, we only care about local replacements.

File: tests/unit/data/sboms/merged.bom.json
Comment (chmeliik) [reply tests/unit/data/sboms/merged.bom.json]: Taking the higher of the two specVersions would probably be fine, but again, I don't really need it in a temporary solution

Comment (brunoapimentel) [reply tests/unit/data/sboms/merged.bom.json]: Hopefully we will have a better solution for multiple SBOMs by then.

File: utils/merge_syft_sbom.py
Comment (chmeliik) [reply utils/merge_syft_sbom.py:31]: Maybe we just check if the version is any url by checking if there's a `:`? (Same logic as https://github.com/containerbuildsystem/cachi2/pull/229#discussion_r1259464017)

Comment (brunoapimentel) [reply utils/merge_syft_sbom.py:31]: > Maybe we just check if the version is any url by checking if there's a :
Yeah, that might do it :)

Comment (brunoapimentel) [reply utils/merge_syft_sbom.py:31]: I'll double check the behavior for v3 lockfiles.

Comment (brunoapimentel) [reply utils/merge_syft_sbom.py:31]: Syft does use only the version for lockfiles v2/v3 :facepalm: 

So this means we need to either:
- fallback to the name-only deduping method we are using for Pip
- live with the duplicates
- update Cachi2 so it also outputs the version for file/vcs dependencies

File: utils/merge_syft_sbom.py
Comment (lkolacek) [inline utils/merge_syft_sbom.py:118]: How the component is passed to function `component_is_duplicated(component: dict[str, Any])` here?

**Review by lkolacek — APPROVED** [review]
As far as I can tell looks great! :+1: Left just 1 minor to educate myself.

File: utils/merge_syft_sbom.py
Comment (brunoapimentel) [reply utils/merge_syft_sbom.py:118]: It is actually used here: 
https://github.com/containerbuildsystem/cachi2/blob/ea97c3dbada30ef0527bf4664a72350c749e55a9/utils/merge_syft_sbom.py#L129-L133

The catch here is that the `component_is_duplicated` function that gets generated holds the info on the existing Cachi2 non-registry [components](https://github.com/containerbuildsystem/cachi2/blob/ea97c3dbada30ef0527bf4664a72350c749e55a9/utils/merge_syft_sbom.py#L96).

**Review by ben-alkov — COMMENTED** [review]
LGTM, but I'd rather Adam have a last look at it before final approval and merge

File: utils/merge_syft_sbom.py
Comment (lkolacek) [reply utils/merge_syft_sbom.py:118]: Aha thanks :)

File: utils/merge_syft_sbom.py
Comment (brunoapimentel) [reply utils/merge_syft_sbom.py]: Updating this thread here: we have decided to use the same rule to dedup non-registry components for both NPM and Pip (that is, check Cachi2 reported non-registry components, and remove all Syft components that have a matching name). So, yeah, it makes sense to use `vcs_url | download_url` qualifiers.

**Comment by brunoapimentel** [comment]
Just FYI: there's apparently a Syft bug that makes it report Pip dependencies twice, adding `%20\\` to the second one ([e.g.](https://github.com/brunoapimentel/rhtap-cachi2-sbom-merge/blob/main/pip/cachi2/rhtap.bom.json#L2074)).

I ignored it for now, since it seems to be fixed in the newer Syft versions.

**Review by taylormadore — APPROVED** [review]
As far as I can tell (I'm not too familiar with Syft), this LGTM

**Pull request merged**
